src/Orderings.jl

```diff
@@ -1,37 +1,29 @@
-module Orderings
-
 import Base.Order: lt, Ordering
-export LexicographicOrdering, LexicographicOrder, LenLexOrder
-
-abstract type LexicographicOrdering <: Ordering end
-
-
-## Assuming the order is given by dictionary with entries (::"letter", ::int)
-struct LexicographicOrder <: LexicographicOrdering
-    order::Dict
-end
+export LenLexOrder
 
+"""
+    struct LenLexOrder{T<:Alphabet} <: Ordering
 
-struct LenLexOrder{LexOrd<:LexicographicOrdering} <: Ordering
-    lexord::LexOrd
+Basic structure representing Length+Lexicographic (left-to-right) ordering of
+the words over given Alphabet. Lexicographing ordering of an Alphabet is
+implicitly specified inside Alphabet struct.
+"""
+struct LenLexOrder{T<:Alphabet} <: Ordering
+    A::T
 end
 
-# `T` shall be later changed to the type of letters of the Alphabet (Char?)
-lt(o::LexicographicOrder, a::T, b::T) = isless(o.order[a], o.order[b])
-
+"""
+    lt(o::LenLexOrder, p::T, q::T) where T<:AbstractWord
 
-## Assuming the following methods are implemented for type Word:
-## length, iteration
-## Assuming that for "letters" from Alphabet exists method: ==
+Return whether the first word is less then the other one in a given LenLex ordering.
+"""
 function lt(o::LenLexOrder, p::T, q::T) where T<:AbstractWord
     if length(p) == length(q)
-        for i in 1:length(p)
-            p[i]==q[i] ||  return lt(o.lexord, p[i], q[i])
+        for (a, b) in zip(p,q)
+            o.A[a] == o.A[b] || return isless(o.A[o.A[a]], o.A[o.A[b]])  # comparing only on positive pointer values
         end
         return false
     else
         return isless(length(p), length(q))
     end
 end
```

Comment:
For me, there are a few details omitted here.

- The assumption that `T <: AbstractWord` is too weak for the proposed implementation. How is this function going to behave for `Word{String}`? In line 23 you would call the alphabet using a string as the index which may not be valid if `A` is not the alphabet of strings. We need to get rid of either the generality, or the index calls.
- In line 23 you seek for a symbol in the alphabet, which is unnecessary. The symbols in the alphabet are equal if and only if `a` and `b` are equal. For the same reason, the call to `isless` is overcomplicated, because `A[A[x]] = x` if `x` is a valid index (or symbol). So, what I would suggest is
```julia
function lt(o::LenLexOrder, p::T, q::T) where T <: AbstractWord
    if length(p) == length(q)
        for (a, b) in zip(p,q)
            a == b || return isless(a, b) 
        end
        return false
    else
        return isless(length(p), length(q))
    end
end
```
but then, why do we even care what the alphabet is if we use numeric indices?

Reply:
for `LenLex` we actually don't care about the particular Alphabet.

My intention of implementing two `Base.getindex(A::Alphabet, x)` was precisely to make `A[A[i]] == i` for every valid index.

Reply:
I believe we now run into certain problems with `Word` struct.

- I believe assumption is not too weak. What is too weak is the line 23 which should be of the form
`o.A[o.A[a]] == o.A[o.A[b]] || return isless(o.A[o.A[a]], o.A[o.A[b]])`
so that we will always compare pointers (they - i.e. positive pointers - store the implicit lexicographic order and iteration in `Word` is an iteration over the array of pointers).
- > the call to isless is overcomplicated, because `A[A[x]] = x` if `x`
Now that is not true:
```
A = KnuthBendix.Alphabet(["a", "b", "c", "d"])
KnuthBendix.set_inversion!(A, "a", "b")

julia> A[A[-1]]
2
```
and this is precisely why I used that weird construction (not seeing anything more sensible): `-1 == 2` will not work in the desired way (`isless` too). Which is - incidentally - a problem with the `==` method for words. I just checked and
```
u1 = KnuthBendix.Word([1,2])
u2 = KnuthBendix.Word([-2,-1])

julia> u1==u2
false
```
and should be `true` (the alphabet as above).

- > but then, why do we even care what the alphabet is if we use numeric indices?

I believe its role is to combine introduction of lexicographic order, inverses and - in a future - pretty printing of words.

Reply:
Okay, now I see why you used such an approach and it makes sense. But then, we need to assume that `T <: Word{Integer}`. Otherwise, the function will not work properly in general.

Reply:
very well, you're correct! ;)
I had the simplistic view of the unique representation of a letter as a pointer, but that is incorrect.

So now we have a situation where words can be compared only in the context of an Alphabet.
hmm... that is less than ideal, but maybe I need to wrap my head around it;

as of `o.A[o.A[a]]` it does what it should (given that letters of Alphabet aren't integers), it's just not what I assumed it will :D

Reply:
maybe my "smart" idea of using negative numbers as pointers to inverses wasn't that smart after all: without it we have a unique representation of each word...

Reply:
> But then, we need to assume that `T <: Word{Integer}`. Otherwise, the function will not work properly in general.

Yes, I see it now.

Reply:
I checked it, actually like that:
`function lt(o::LenLex, p::T, q::T) where T<:AbstractWord{<:Integer}`

Reply:
> > But then, we need to assume that `T <: Word{Integer}`. Otherwise, the function will not work properly in general.
> 
> Yes, I see it now.

This was my silent assumption as well, so we can either enforce it here (as you did), or make 
```julia
abstract type AbstractWord{T<:Integer} <: AbstractVector{T} end
```